src/ILCompiler.Compiler/src/Compiler/DelegateCreationInfo.cs

```diff
@@ -0,0 +1,131 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+using Internal.IL;
+using Internal.IL.Stubs;
+using Internal.TypeSystem;
+using ILCompiler.DependencyAnalysis;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler
+{
+    /// <summary>
+    /// Captures information required to generate a ReadyToRun helper to create a delegate type instance
+    /// pointing to a specific target method.
+    /// </summary>
+    public sealed class DelegateCreationInfo
+    {
+        /// <summary>
+        /// Gets the node corresponding to the method that initializes the delegate.
+        /// </summary>
+        public IMethodNode Constructor
+        {
+            get; private set;
+        }
+
+        /// <summary>
+        /// Gets the node representing the target method of the delegate.
+        /// </summary>
+        public ISymbolNode Target
+        {
+            get; private set;
+        }
+
+        /// <summary>
+        /// Gets an optional node passed as an additional argument to the constructor.
+        /// </summary>
+        public IMethodNode Thunk
+        {
+            get; private set;
+        }
+
+        private DelegateCreationInfo(IMethodNode constructor, ISymbolNode target, IMethodNode thunk = null)
+        {
+            Constructor = constructor;
+            Target = target;
+            Thunk = thunk;
+        }
+
+        /// <summary>
+        /// Constructs a new instance of <see cref="DelegateCreationInfo"/> set up to construct a delegate of type
+        /// '<paramref name="delegateType"/>' pointing to '<paramref name="targetMethod"/>'.
+        /// </summary>
+        public static DelegateCreationInfo Create(TypeDesc delegateType, MethodDesc targetMethod, NodeFactory factory)
+        {
+            var context = (CompilerTypeSystemContext)delegateType.Context;
+            var systemDelegate = targetMethod.Context.GetWellKnownType(WellKnownType.MulticastDelegate).BaseType;
+
+            int paramCountTargetMethod = targetMethod.Signature.Length;
+            if (!targetMethod.Signature.IsStatic)
+            {
+                // TODO: check for explicit this and don't increment if that's the case
```

Comment:
I think this TODO is not necessary. "explicit this" was really just annotation for managed C++, that runtime does nothing about.